main.tf

```diff
@@ -55,6 +55,13 @@ resource "azurerm_network_security_rule" "predefined_rules_for" {
   source_application_security_group_ids      = lookup(each.value, "source_application_security_group_ids", null)
   source_port_range                          = lookup(each.value, "source_port_range", "*") == "*" ? "*" : null
   source_port_ranges                         = lookup(each.value, "source_port_range", "*") == "*" ? null : split(",", each.value.source_port_range)
+
+  lifecycle {
+    precondition {
+      condition     = contains(keys(each.value), "priority")
```

Comment:
I've tried the following code in example:

```hcl
predefined_rules = [
    {
      name     = "Cassandra"
      priority = null
    },
  ]
```

The error message is:

```text
Error: Missing required argument
│ 
│   with module.nsg.module.nsg.azurerm_network_security_rule.predefined_rules_for["Cassandra"],
│   on ../../main.tf line 45, in resource "azurerm_network_security_rule" "predefined_rules_for":
│   45:   priority                                   = each.value.priority
│ 
│ The argument "priority" is required, but no definition was found.
```

The `priority` key exists with `null` value could bypass your `precondition`, so I would recommend:

```hcl
condition     = try(each.value.priority >= 100 && each.value.priority <= 4096, false)
```

Which would response the following error message to `null` priority:

```text
Error: Resource precondition failed
│ 
│   on ../../main.tf line 61, in resource "azurerm_network_security_rule" "predefined_rules_for":
│   61:       condition     = try(each.value.priority >= 100 && each.value.priority <= 4096, false)
│     ├────────────────
│     │ each.value.priority is null
│ 
│ Precondition failed: 'predefined_rules.priority' must be provided for predefined rules if 'var.use_for_each' is set to
│ true.
```

We can also tell our users that the priority must between [100-4096](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/network_security_rule#priority).

Reply:
Good catch, modified the precondition with you suggestion.